File: backend/KtorApp/src/main/kotlin/com/example/utils/DatabaseFactory.kt
package com.example.utils

import com.mongodb.ConnectionString
import com.mongodb.MongoClientSettings
import com.mongodb.client.MongoClient
import com.mongodb.client.MongoClients
import com.mongodb.client.MongoCollection
import com.mongodb.client.MongoDatabase
import org.bson.Document
import org.slf4j.Logger
import org.slf4j.LoggerFactory

object DatabaseFactory {
    private val logger: Logger = LoggerFactory.getLogger(DatabaseFactory::class.java)

    private lateinit var mongoClient: MongoClient
    private lateinit var database: MongoDatabase

    /**
     * Инициализирует подключение к MongoDB.
     * Читает параметры подключения из переменных окружения.
     */
    fun init() {
        try {
            val connectionString = System.getenv("MONGODB_URI") ?: "mongodb://mongo:27018/bookdb"
            val dbName = "bookdb" // Или можно извлечь из строки подключения

            logger.info("Инициализация подключения к MongoDB с строкой подключения: $connectionString")
            val settings = MongoClientSettings.builder()
                .applyConnectionString(ConnectionString(connectionString))
                .build()

            mongoClient = MongoClients.create(settings)
            database = mongoClient.getDatabase(dbName)
            logger.info("Подключение к MongoDB успешно установлено. Используемая база данных: $dbName")
        } catch (e: Exception) {
            logger.error("Ошибка при подключении к MongoDB: ${e.message}", e)
            throw e // Перебрасываем исключение, чтобы приложение не запускалось без подключения к БД
        }
    }

    /**
     * Получает коллекцию пользователей.
     * @return Коллекция `users`.
     */
    fun getUsersCollection(): MongoCollection<Document> {
        ensureInitialized()
        return database.getCollection("users")
    }

    /**
     * Получает коллекцию книг.
     * @return Коллекция `books`.
     */
    fun getBooksCollection(): MongoCollection<Document> {
        ensureInitialized()
        return database.getCollection("books")
    }

    /**
     * Получает коллекцию текстовых блоков.
     * @return Коллекция `text_blocks`.
     */
    fun getTextBlocksCollection(): MongoCollection<Document> {
        ensureInitialized()
        return database.getCollection("text_blocks")
    }

    /**
     * Закрывает подключение к MongoDB.
     */
    fun close() {
        if (::mongoClient.isInitialized) {
            logger.info("Закрытие подключения к MongoDB...")
            mongoClient.close()
            logger.info("Подключение к MongoDB закрыто.")
        }
    }

    /**
     * Проверяет, инициализирована ли база данных.
     * @throws IllegalStateException Если база данных не инициализирована.
     */
    private fun ensureInitialized() {
        if (!::mongoClient.isInitialized || !::database.isInitialized) {
            throw IllegalStateException("DatabaseFactory не инициализирован. Вызовите DatabaseFactory.init() перед использованием.")
        }
    }
}

File: backend/utils/DatabaseFactory.py
import os
import logging
from pymongo import MongoClient, errors

logger = logging.getLogger(__name__)

class DatabaseFactory:
    _mongo_client = None
    _database = None

    @classmethod
    def init(cls):
        try:
            uri = os.environ.get("MONGODB_URI", "mongodb://mongo:27018/bookdb")
            db_name = "bookdb" # Or extract from connection string if needed.

            logger.info(f"Initializing MongoDB connection with URI: {uri}")
            cls._mongo_client = MongoClient(uri)
            cls._database = cls._mongo_client[db_name]
            logger.info(f"MongoDB connection established successfully. Database used: {db_name}")

        except errors.ConnectionFailure as e:
            logger.error(f"Error connecting to MongoDB: {e}")
            raise  # Re-raise the exception to halt application startup.


    @classmethod
    def getUsersCollection(cls):
        cls._ensure_initialized()
        return cls._database["users"]

    @classmethod
    def getBooksCollection(cls):
        cls._ensure_initialized()
        return cls._database["books"]

    @classmethod
    def getTextBlocksCollection(cls):
        cls._ensure_initialized()
        return cls._database["text_blocks"]

    @classmethod
    def close(cls):
        if cls._mongo_client:
            logger.info("Closing MongoDB connection...")
            cls._mongo_client.close()
            logger.info("MongoDB connection closed.")

    @classmethod
    def _ensure_initialized(cls):
        if not cls._mongo_client or not cls._database:
            raise RuntimeError("DatabaseFactory not initialized. Call DatabaseFactory.init() before use.")


#Example usage (Remember to configure logging appropriately):
#logging.basicConfig(level=logging.INFO)
#DatabaseFactory.init()
#users_collection = DatabaseFactory.getUsersCollection()
#DatabaseFactory.close()